src/linear_algebra/tensor_product.lean

```diff
@@ -366,6 +366,30 @@ begin
   { intros t₁ t₂ ht₁ ht₂, exact submodule.add_mem _ ht₁ ht₂, },
 end
 
+variables {R M N}
+
+theorem span_tmul_set_eq_top {sM : set M} {sN : set N} (hM : submodule.span R sM = ⊤)
+  (hN : submodule.span R sN = ⊤) :
+  submodule.span R { t : M ⊗[R] N | ∃ (m ∈ sM) (n ∈ sN), m ⊗ₜ n = t } = ⊤ :=
+begin
+  ext t, simp only [submodule.mem_top, iff_true],
+  apply t.induction_on,
+  { exact submodule.zero_mem _, },
+  { intros m' n',
+    have hm' : m' ∈ submodule.span R sM := by {rw hM, exact submodule.mem_top},
+    have hn' : n' ∈ submodule.span R sN := by {rw hN, exact submodule.mem_top},
+    refine submodule.span_induction hm' _ _ _ _,
+    { intros m hm, refine submodule.span_induction hn' _ _ _ _,
+      { intros n hn, apply submodule.subset_span, use [m, hm, n, hn], },
+      { rw tmul_zero, exact submodule.zero_mem _, },
+      { intros n₁ n₂ hn₁ hn₂, rw tmul_add, exact submodule.add_mem _ hn₁ hn₂, },
+      { intros r n hn, rw tmul_smul, exact submodule.smul_mem _ r hn, }, },
+    { rw zero_tmul, exact submodule.zero_mem _, },
+    { intros m₁ m₂ hm₁ hm₂, rw add_tmul, exact submodule.add_mem _ hm₁ hm₂, },
+    { intros r m hm, rw ←smul_tmul', exact submodule.smul_mem _ r hm, }, },
+  { intros t₁ t₂ ht₁ ht₂, exact submodule.add_mem _ ht₁ ht₂, },
+end
```

Comment:
golf:
```suggestion
begin
  apply le_antisymm, { intros _ _, triv },
  rw [← span_tmul_eq_top, submodule.span_le], rintro t ⟨m, n, rfl⟩, rw set_like.mem_coe,
  apply submodule.span_induction (hM.substr (by triv : m ∈ ⊤)),
  { intros m hm, apply submodule.span_induction (hN.substr (by triv : n ∈ ⊤)),
    { exact λ n hn, submodule.subset_span ⟨m, hm, n, hn, rfl⟩, },
    { rw tmul_zero, exact submodule.zero_mem _, },
    { intros n₁ n₂ hn₁ hn₂, rw tmul_add, exact submodule.add_mem _ hn₁ hn₂, },
    { intros r n hn, rw tmul_smul, exact submodule.smul_mem _ r hn, }, },
  { rw zero_tmul, exact submodule.zero_mem _, },
  { intros m₁ m₂ hm₁ hm₂, rw add_tmul, exact submodule.add_mem _ hm₁ hm₂, },
  { intros r m hm, rw ←smul_tmul', exact submodule.smul_mem _ r hm, },
end
```

I think this proof should be preferred if [my two lemmas](https://github.com/leanprover-community/mathlib/compare/tensor_fin_dim?expand=1#diff-0f91a25ed5b4100fb1b118ded829faef51d44739010fda07b70055f7822adfdaR572) are not worth having; otherwise my proof should be preferred. What do you think?

Reply:
I like the look of your proof @alreadydone; although I wonder if we should define a `submodule.tmul` function that takes two submodules to the submodule of the tensor product generated by them, and then state your lemmas in terms of that.

Reply:
Concretely:
```lean
def submodule.tmul (p : submodule R M) (q : submodule R N) : submodule R (M ⊗[R] N) :=
⨆ (m : ↥p), q.map $ tensor_product.mk R M N m
```
which matches how we define `submodule.has_mul`

Reply:
If we do that, we can then state things like `submodule.fg.tmul (hp : p.fg) (hq : q.fg) : (p.tmul q).fg` (assuming that's true), which is a more general result than the one in this PR.

Reply:
I have a PR I'll push shortly that generalizes some of the `*` machinery

Reply:
Done in #13709